auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java

```diff
@@ -61,38 +64,59 @@ public void startSignIn(@NonNull final String email,
                     .build();
         }
 
-        // Kick off the flow including signing in, linking accounts, and saving with SmartLock
-        getAuth().signInWithEmailAndPassword(email, password)
-                .continueWithTask(new Continuation<AuthResult, Task<AuthResult>>() {
-                    @Override
-                    public Task<AuthResult> then(@NonNull Task<AuthResult> task) throws Exception {
-                        // Forward task failure by asking for result
-                        AuthResult result = task.getResult(Exception.class);
+        // We first check if we are trying to upgrade an anonymous user
+        // Calling linkWithCredential will fail since the email exists, so we can save an RPC call
+        // We just need to validate the credential
+        if (AnonymousUpgradeUtils.canUpgradeAnonymous(getAuth(), getArguments())) {
+            final AuthCredential credToValidate = EmailAuthProvider.getCredential(email, password);
+            AnonymousUpgradeUtils.validateCredential(credToValidate).addOnCompleteListener(
+                    new OnCompleteListener<AuthResult>() {
+                        @Override
+                        public void onComplete(@NonNull Task<AuthResult> task) {
+                            if (task.isSuccessful()) {
+                                handleMergeFailure(credToValidate);
+                            } else {
+                                setResult(Resource.<IdpResponse>forFailure(task.getException()));
+                            }
+                        }
+                    });
+
+        } else {
```

Comment:
Same nit as before.

When a method is just

```
if (condition) {
   action1;
} else {
   action2;
}
```

Consider:

```
if (condition) {
   action1;
   return;
}

action2
```

With something like you have here it helps from getting into crazy indentation levels.

Reply:
Done.